File: Java/118.PascalsTriangle.java
/**
* Given a non-negative integer numRows, generate the first numRows of Pascal's triangle.
* 
* Example of the first 5 rows in a Pascal's triangle
*
*     1
*    1 1
*   1 2 1
*  1 3 3 1
* 1 4 6 4 1
* 
*/

class Solution {
    public List<List<Integer>> generate(int numRows) {
        
        List<List<Integer>> triangle = new ArrayList<>();
        if (numRows == 0) {
            // for the case when the numRows is zero, return an empty list
            return triangle;
        }
        
        initializePascalTriangle(triangle);
        
        // For every row
        for (int i = 1; i < numRows; i++) {
            List<Integer> line = new ArrayList<>();
            // every line starts with a 1
            line.add(1);
            
            // Fill the row
            for(int j = 1; j < i; j++) {
                int element = triangle.get(i-1).get(j) +
                              triangle.get(i-1).get(j-1);
                line.add(element);
            }
            
            // every line ends with a 1
            line.add(1);
            // add line to the triangle
            triangle.add(line);
        }
        
        // Return the filled triangle
        return triangle;
    }
    
    /**
    * Initialize the triangle object and add the first element
    */
    private void initializePascalTriangle(List<List<Integer>> triangle) {
        // Initialize a list of list
        triangle.add(new ArrayList<>());
        // Add the first element of the triangle
        triangle.get(0).add(1);
    }
}

File: Python/118. PascalsTriangle.py
# a function that receives a number a will create
# a pascal triangle of that height


class Solution(object):

    def generate(self, num):
        """
        creates a two dimensional array holding a pascal triangle of
        the desired height
        :param num: int the height of the triangle
        :return: List[List[int]] a matrix of the pascal triangle
        """
        
        # when the num is zero, return an empty list
        if num == 0:
            return []

        # the base of the triangle
        return_triangle = [[1]]

        # keeps adding a layer
        for i in range(1, num):

            # creates a new line
            new_line = []

            # beginning of line is always 1
            new_line.append(1)

            # middle of line gets filled by upper level
            for j in range(1, i):
                new_line.append(return_triangle[i - 1][j] +
                                return_triangle[i - 1][j - 1])

            # end of line
            new_line.append(1)

            # adds the line
            return_triangle.append(new_line)

        # returns the triangle
        return return_triangle
